presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java

```diff
@@ -0,0 +1,471 @@
+                .add("taskId", taskId)
+                .add("remainingDrivers", remainingDrivers.get())
+                .toString();
+    }
+
+    private class DriverSplitRunnerFactory
+    {
+        private final DriverFactory driverFactory;
+        private final PipelineContext pipelineContext;
+
+        private final AtomicInteger pendingCreation = new AtomicInteger();
+        private final AtomicBoolean noMoreDriverRunner = new AtomicBoolean();
+        private boolean closed;
+
+        private DriverSplitRunnerFactory(DriverFactory driverFactory, boolean partitioned)
+        {
+            this.driverFactory = requireNonNull(driverFactory, "driverFactory is null");
+            this.pipelineContext = taskContext.addPipelineContext(driverFactory.getPipelineId(), driverFactory.isInputDriver(), driverFactory.isOutputDriver(), partitioned);
+        }
+
+        public DriverSplitRunner createDriverRunner(@Nullable ScheduledSplit partitionedSplit)
+        {
+            checkState(!noMoreDriverRunner.get(), "Cannot create driver for pipeline: %s", pipelineContext.getPipelineId());
+            pendingCreation.incrementAndGet();
+            // create driver context immediately so the driver existence is recorded in the stats
+            // the number of drivers is used to balance work across nodes
+            DriverContext driverContext = pipelineContext.addDriverContext(Lifespan.taskWide());
+            return new DriverSplitRunner(this, driverContext, partitionedSplit);
+        }
+
+        public Driver createDriver(DriverContext driverContext, @Nullable ScheduledSplit partitionedSplit)
+        {
+            Driver driver = driverFactory.createDriver(driverContext);
+
+            if (partitionedSplit != null) {
+                // TableScanOperator requires partitioned split to be added before the first call to process
+                driver.updateSource(new TaskSource(partitionedSplit.getPlanNodeId(), ImmutableSet.of(partitionedSplit), true));
+            }
+
+            verify(pendingCreation.get() > 0, "pendingCreation is expected to be greater than zero");
+            pendingCreation.decrementAndGet();
+
+            closeDriverFactoryIfFullyCreated();
+
+            return driver;
+        }
+
+        public void noMoreDriverRunner()
+        {
+            if (noMoreDriverRunner.get()) {
+                return;
+            }
+            noMoreDriverRunner.set(true);
+            closeDriverFactoryIfFullyCreated();
+        }
+
+        public boolean isNoMoreDriverRunner()
+        {
+            return noMoreDriverRunner.get();
+        }
+
+        public void closeDriverFactoryIfFullyCreated()
+        {
+            if (closed) {
+                return;
+            }
+            if (isNoMoreDriverRunner() && pendingCreation.get() == 0) {
+                driverFactory.noMoreDrivers(Lifespan.taskWide());
+                driverFactory.noMoreDrivers();
+                closed = true;
```

Comment:
Curious: why need to track `closed` in `DriverSplitRunnerFactory`?

Reply:
Probably to avoid calling `noMoreDrivers` more than once. I moved it from the `SqlTaskExecution`. I wonder if some  `DriferFactory` can rely on that. I just realized that having a simple boolean might not be thread safe. Replaced with `AtomicBoolean`

Reply:
@arhimondr I see. This is added in the grouped execution commit : 4e55aadca50e2b6f904fcfa444540e90c02e3383 . `closed` doesn't exist prior to it:

https://github.com/prestodb/presto/blob/0309ca4d30aa281b55ca526fda74b946dd3a805f/presto-main/src/main/java/com/facebook/presto/execution/SqlTaskExecution.java#L534-L539

So probably a safe mechanism added for grouped execution, since `noMoreDrivers` now need to be called both for lifespan and overall...